cpp/ql/test/experimental/query-tests/Security/CWE/CWE-754/semmle/tests/test.cpp

```diff
@@ -0,0 +1,101 @@
+int scanf(const char *format, ...);
+int globalVal;
+char * globalVala;
+int * globalValp;
+char globalVala2;
+int functionWork1() {
+  int i;
+  char a[10];
+  int b;
+  int *p = &b;
+  if (scanf("%i", &i) != 1) // GOOD
+    return -1;
+  if (scanf("%s", a) != 1) // GOOD
+    return -1;
+  if (scanf("%i", p) != 1) // GOOD
+    return -1;
+  return i;
+}
+
+int functionWork1_() {
+  int i;
+  char a[10];
+  int b;
+  int *p = &b;
+  int r;
+  r = scanf("%i", &i);
+  if (r != 1) // GOOD
+    return -1;
+  r = scanf("%s", a);
+  if (r == 1) // GOOD
+    return -1;
+  r = scanf("%i", p);
+  if (r != 1) // GOOD
+    return -1;
+  return i;
+}
+
+int functionWork1b() {
+  int i;
+  char a[10];
+  int b;
+  int *p = &b;
+  scanf("%i", &i); // BAD
+  scanf("%s", a); // BAD
+  scanf("%i", p); // BAD
+  return i;
+}
+
+int functionWork2() {
+  int i = 0;
+  char a[10] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+  int b = 1;
+  int *p = &b;
+  scanf("%i", &i); // GOOD:the error can be determined by examining the initial value.
```

Comment:
```
// GOOD:the error can be determined by examining the initial value.
```
Technically this is not quite the case, as one of two things might have happened:

1. `scanf` succeeded and read a `0` value.
2. `scanf` failed, but this is inconsequential, because `i` was initialized.

So maybe write something like: "Argument initialized even when `scanf` fails"